File: eform-client/src/app/plugins/modules/backend-configuration-pn/models/calendar/calendar-task-request.model.ts
import {CalendarRepeatMeta, CalendarRepeatRule} from './calendar-task.model';

export interface CalendarTaskCreateModel {
  title: string;
  startHour: number;
  duration: number;
  tags: string[];
  assigneeIds: number[];
  boardId: number;
  color: string;
  descriptionHtml: string;
  repeatRule: CalendarRepeatRule;
  repeatMeta?: CalendarRepeatMeta;
  taskDate: string;
  driveLink?: string;
  propertyId: number;
  // Selected eForm (SDK CheckList id), null when the event has no eForm.
  // Declared explicitly so the field is compiler-checked: it used to reach the
  // backend only because the modal typed its payload `any`, which made a
  // rename/typo silently drop the eForm from the request.
  eformId: number | null;
  // CSV of JS getDay() weekday indices ("1,3,5") — only populated for
  // multi-day weekly custom rules. Cleared (sent as null) for any non-custom
  // rule so the backend column is wiped on rule change.
  repeatWeekdaysCsv?: string | null;
}

export interface CalendarTaskUpdateModel extends CalendarTaskCreateModel {
  id: number;
  repeatSeriesId?: string;
  // Pre-edit occurrence date ("YYYY-MM-DD") for scope-aware edits (#885).
  originalDate?: string;
}

export type RepeatEditScope = 'this' | 'thisAndFollowing' | 'all';
export type RepeatDeleteScope = 'this' | 'thisAndFollowing' | 'thisAndFollowingIncludingCompleted' | 'all';

/**
 * The POST body of `calendar/tasks/week` — the grid's only read path, shared by
 * the week/day/schedule load and by each of the six calls the month view's
 * forkJoin makes.
 *
 * An object rather than a positional argument list: the server model
 * (`CalendarTaskRequestModel`) has grown a filter per epic, and the last
 * positional form already ended in three same-typed collections, where a
 * transposed pair compiles cleanly and silently filters by the wrong thing.
 *
 * `siteIds` and `workerTagIds` are two assignee filters the server ORs (#1212),
 * but they are NOT symmetric. Before matching, the server widens the tag set to
 * `workerTagIds` plus every worker tag the selected `siteIds` belong to; a task
 * survives when its explicit assignees intersect `siteIds` OR its assigned
 * worker tags intersect that widened set.
 *
 * The widening runs one way only, sites → tags — there is no tags → sites
 * expansion. So picking an individual ALSO matches tasks assigned to a team
 * that individual is in, while picking a team does NOT match that team's
 * members' own individually-assigned events. An empty `workerTagIds` therefore
 * does not mean "no tag matching" whenever `siteIds` is non-empty.
 *
 * The lists are never intersected, and the client never post-filters —
 * everything the grid narrows by travels on this model.
 *
 * `tagNames` is the planning/eForm tag filter and has nothing to do with
 * `workerTagIds`; see the note on the server model of the same name.
 */
export interface CalendarTaskWeekRequestModel {
  propertyId: number;
  weekStart: string;
  weekEnd: string;
  boardIds: number[];
  tagNames: string[];
  /** Explicit assignee (site) ids. Empty = no narrowing by individual. */
  siteIds: number[];
  /** Worker-tag ("team") ids. Empty = no narrowing by team. */
  workerTagIds: number[];
}

File: eFormAPI/Plugins/BackendConfiguration.Pn/BackendConfiguration.Pn/Infrastructure/Models/Calendar/CalendarTaskRequestModel.cs
using System.Collections.Generic;

namespace BackendConfiguration.Pn.Infrastructure.Models.Calendar;

public class CalendarTaskRequestModel
{
    public int PropertyId { get; set; }
    public string WeekStart { get; set; }
    public string WeekEnd { get; set; }
    public List<int> BoardIds { get; set; } = [];
    public List<string> TagNames { get; set; } = [];
    public List<int> SiteIds { get; set; } = [];

    /// <summary>
    /// Optional worker-tag ("team") filter (#1212). Combines with
    /// <see cref="SiteIds"/> by OR, never AND — but the two are <em>not</em>
    /// symmetric. Before matching, the server widens this list into the
    /// <em>effective</em> tag set: this list plus every worker tag the requested
    /// <see cref="SiteIds"/> are members of. A task then survives the assignee
    /// filter when its explicit assignees intersect <see cref="SiteIds"/>
    /// <em>or</em> its assigned worker tags intersect that effective set.
    ///
    /// The widening runs one way only, sites → tags; there is no tags → sites
    /// expansion anywhere in the filter path. So selecting an individual
    /// <em>does</em> match tasks assigned to a team that individual belongs to,
    /// while selecting a team does <em>not</em> match that team's members' own
    /// individually-assigned tasks. (<c>CalendarTaskResponseModel.TeamAssigneeIds</c>
    /// is populated separately and never read by the filter — team membership
    /// matches only through the effective tag set.) Empty here is therefore not
    /// the same as "no tag matching": with <see cref="SiteIds"/> set, the
    /// effective set is still non-empty. Both lists empty = no assignee
    /// filtering at all.
    ///
    /// Not to be confused with <see cref="TagNames"/>, which is the
    /// planning/eForm tag filter — a different concept entirely.
    /// </summary>
    public List<int> WorkerTagIds { get; set; } = [];

    /// <summary>
    /// When true, the calendar emits only *actionable* compliance rows for the requested
    /// week — i.e. compliances whose backing SDK Case still exists, is not soft-deleted,
    /// and is not yet completed (Status != 100). This is intended for the mobile-worker
    /// gRPC path (<c>EventsGrpcService</c>) where non-actionable rows have no write
    /// handler to bind to and would just clutter the worker's view.
    ///
    /// Default <c>false</c> preserves the historical behavior used by the angular admin
    /// calendar (<c>CalendarController</c>) and other gRPC consumers
    /// (<c>CalendarGrpcService</c>): all in-week compliances surface, including missed
    /// and completed ones, so the admin can audit the full week.
    /// </summary>
    public bool ActionableOnly { get; set; } = false;

    /// <summary>
    /// Optional explicit language for Title/Description selection. Set by the
    /// mobile-worker gRPC path (<c>EventsGrpcService</c>) to the worker's
    /// SDK <c>Site.LanguageId</c> so the worker sees the task in their own
    /// language. Null (angular admin REST / <c>CalendarGrpcService</c>) → the
    /// current user's language, preserving existing behaviour.
    /// </summary>
    public int? LanguageId { get; set; }
}
